File: ios/src/main/java/y2k/joyreactor/platform/PlatformImpl.java
package y2k.joyreactor.platform;

import kotlin.Unit;
import org.robovm.apple.foundation.*;
import org.robovm.apple.uikit.UIImage;
import rx.Observable;
import y2k.joyreactor.common.ObservableExtensionsKt;

import java.io.File;
import java.util.List;

/**
 * Created by y2k on 29/09/15.
 */
public class PlatformImpl extends Platform {

    private static final File CURRENT_DIRECTORY = dogGetCurrentDirectory();

    private static File dogGetCurrentDirectory() {
        List<String> dirs = NSPathUtilities.getSearchPathForDirectoriesInDomains(
            NSSearchPathDirectory.DocumentDirectory,
            NSSearchPathDomainMask.UserDomainMask, true);
        return new File(dirs.get(0));
    }

    @Override
    public File getCurrentDirectory() {
        return CURRENT_DIRECTORY;
    }

    @Override
    public Navigation getNavigator() {
        return new StoryboardNavigation();
    }

    @Override
    public byte[] loadFromBundle(String name, String ext) {
        String path = NSBundle.getMainBundle().findResourcePath(name, ext);
        return NSData.read(new File(path)).getBytes();
    }

    @Override
    public Observable<?> saveToGallery(File imageFile) {
        return ObservableExtensionsKt.ioUnitObservable(() -> {
            new UIImage(imageFile).saveToPhotosAlbum(null);
            return Unit.INSTANCE;
        });
    }
}

File: ios/src/main/kotlin/y2k/joyreactor/platform/PlatformImpl.kt
package y2k.joyreactor.platform

import com.j256.ormlite.support.ConnectionSource
import org.robovm.apple.foundation.*
import org.robovm.apple.uikit.UIImage
import rx.Observable
import y2k.joyreactor.common.ioUnitObservable
import y2k.joyreactor.common.platform.NavigationService
import y2k.joyreactor.common.platform.Platform
import java.io.File

/**
 * Created by y2k on 29/09/15.
 */
class PlatformImpl : Platform {

    override fun <T> decodeImage(path: File): T {
        TODO()
    }

    override fun buildConnection(file: File): ConnectionSource {
        throw UnsupportedOperationException()
    }

    override val currentDirectory: File
        get() = CURRENT_DIRECTORY

    override val navigator: NavigationService
        get() = StoryboardNavigation()

    override fun loadFromBundle(name: String, ext: String): ByteArray {
        val path = NSBundle.getMainBundle().findResourcePath(name, ext)
        return NSData.read(File(path)).bytes
    }

    override fun saveToGallery(imageFile: File): Observable<*> {
        return ioUnitObservable {
            UIImage(imageFile).saveToPhotosAlbum(null)
        }
    }

    companion object {

        private val CURRENT_DIRECTORY = dogGetCurrentDirectory()

        private fun dogGetCurrentDirectory(): File {
            val dirs = NSPathUtilities.getSearchPathForDirectoriesInDomains(
                NSSearchPathDirectory.DocumentDirectory,
                NSSearchPathDomainMask.UserDomainMask, true)
            return File(dirs[0])
        }
    }
}